### PR TITLE
Add Icmp header 

### PR DESCRIPTION
**Pull request by scrollins**

### CODE REVIEW

File: test/srv6-sighup-flow/src/nf.rs [hunk 1]
```diff
@@ -33,6 +33,7 @@ pub fn nf<T: 'static + Batch<Header = NullHeader>, S: Scheduler + Sized>(
         .filter(box |pkt| match pkt.get_header().next_header() {
```
Comment (zeeshanlakhani) [inline test/srv6-sighup-flow/src/nf.rs]: I'd probably keep this test as is and create a new one that test's the message

File: framework/src/headers/ip/v6/mod.rs [hunk 2]
```diff
@@ -198,6 +200,7 @@ impl Ipv6Header {
                 // TODO: should validate we are only skipping v6 extension headers
                 while next_hdr != TCP_NXT_HDR
                     && next_hdr != UDP_NXT_HDR
+                    && next_hdr != ICMP_NXT_HDR
```
Comment (zeeshanlakhani) [inline framework/src/headers/ip/v6/mod.rs]: we probably want to do checksum tests specific to icmp as well. Look at packets.rs (in the test folder) as an example. You can extract icmp message bytes from wireshark for example.

File: framework/src/headers/icmp.rs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+use std::default::Default;
+use std::fmt;
+
+pub struct IcmpV6Header<T> {
+    msg_type: u8,
+    code: u8,
+    checksum: u16
+}
+
+pub const ROUTER_ADVERTISEMENT: int = 134;
```
Comment (zeeshanlakhani) [inline framework/src/headers/icmp.rs]: as part of the functional test, we should work on how we become selective on which we accept and which we just want to forward through. Maybe an enum?

**Comment by scrollins** [comment]
@zeeshanlakhani i think this is a good first pass at parsing icmpv6 messages. the functional and unit tests are pretty straight forward. the next pass will be to filter out icmpv6 message types that arent NDP messages. let me know what you think

File: test/icmpv6/src/nf.rs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    dst: Ipv6Addr
+}
+
+impl Default for Meta {
+    fn default() -> Meta {
+        Meta {
+            msg_type: IcmpMessageType::NeighborAdvertisement,
+            code: 0,
+            checksum: 0,
+            src: Ipv6Addr::unspecified(),
```
Comment (zeeshanlakhani) [inline test/icmpv6/src/nf.rs]: do we need these as meta? I know it's kinda a banal exercise, but maybe just having what's in the ipv6 header already works.

File: framework/src/headers/icmp.rs [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+impl<T> EndOffset for IcmpV6Header<T>
+    where
+        T: IpHeader,
+{
+    type PreviousHeader = T;
+
+    #[inline]
+    fn offset(&self) -> usize {
+        54
```
Comment (zeeshanlakhani) [inline framework/src/headers/icmp.rs]: @scrollins, probably want to comment on how we get to this offset here; Probably worth comments on each of these functions too.

File: framework/src/headers/icmp.rs [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    }
+
+    #[inline]
+    fn size() -> usize {
+        32
```
[the remark this answers is not in the document]
Comment (zeeshanlakhani) [reply framework/src/headers/icmp.rs]: Right? I thought that too, dat / 8 right?

[the remark this answers is not in the document]
Comment (drunkirishcoder) [reply framework/src/headers/icmp.rs]: Yep, the unit is `bytes`.

File: test/icmpv6/src/nf.rs (hunk 4 above)
Comment (scrollins) [reply test/icmpv6/src/nf.rs]: i cleaned up the test to only check for the icmpv6 fields that we wanted to compare. i removed the src and dist address in the ipv6 header in the meta struct

File: framework/src/headers/icmp.rs (hunk 5 above)
Comment (scrollins) [reply framework/src/headers/icmp.rs]: i misunderstood and thought the offset was the starting byte in the packet for the icmpv6 header. The offset should be 4. I put comments in the offset and size methods to show how those are calculated. at the top of the icmp header rust src file , i have what the rfc is for icmpv6 messages

File: .gitignore [hunk 7]
```diff
@@ -53,3 +53,5 @@ out.pcap
 
 # wireedit
 *.lnk
+
+.DS_Store
```
[the remark this answers is not in the document]
Comment (scrollins) [reply .gitignore]: done

File: framework/src/headers/icmp.rs [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+    }
+
+    #[inline]
+    fn payload_size(&self, _: usize) -> usize {
+        32
```
[the remark this answers is not in the document]
Comment (scrollins) [reply framework/src/headers/icmp.rs]: done

**Comment by scrollins** [comment]
@zeeshanlakhani lets chat this week about the checksum calc for icmpv6 messages. we can do that work in a follow up PR to this

**Comment by zeeshanlakhani** [comment]
@scrollins in thinking about it, we can punt on checksumming until we advertise ourselves eventually. May be worth putting an issue in the og backlog.

**Pull request merged**